site/content/tutorial/04-logic/05-keyed-each-blocks/text.md

```diff
@@ -2,18 +2,20 @@
 title: Keyed each blocks
 ---
 
-By default, when you modify the value of an `each` block, it will add and remove items at the *end* of the block, and update any values that have changed. That might not be what you want.
+By default, when you add an element to the array that the `each` block iterates over, it will add an item at the *end* of the block instead of at the same position as the element in the array, and update the attributes of the existing items. This means that if you add an element to the array anywhere other than at the end, this leads to each item corresponding to a different element of the array than the one it was created for. The same goes for removing an element from the array. That might not be what you want.
```

Comment:
im thinking whether we can explain this more visually,

pardon my grammar, but my rough idea would be:

---

By default, when you add / remove an element into / from the middle of an array, which the `each` block iterates over, the `each` block has no idea which element was added / removed except that the array has changed.

the `each` block will iterate through every item, and update each item if the props of the component at an index is different from the the data at the index. at the end of the loop, new component will be created if there's more data than the components, or extra components will be removed if there's less data.

Reply:
I agree, that sounds much better. 👍

Also upon thinking about it some more, I think we're currently missing an important thing in the text. This behavior only appears, because components are resused instead of being destroyed and recreated when the underlying array changes.

Also we should probably clarify that this behavior is only relevant for components, since they can hold on to the initial state. For example, it doesn't matter for something like this.

```svelte
{#each array as element}
   <p>Hello {element} world.</p>
{/each}
```